src/crimson/osd/main.cc

```diff
@@ -47,7 +47,7 @@ int main(int argc, char* argv[])
 
   args.insert(begin(args), argv[0]);
   try {
-    return app.run_deprecated(args.size(), const_cast<char**>(args.data()), [&] {
+    return app.run(args.size(), const_cast<char**>(args.data()), [&] {
```

Comment:
we need to *wait* for OSD's shutdown before stopping the engine.

Reply:
app.run will wait program running finish before call engine.exit().  so that is our job to call osd::shutdown explicitly in proper place.

Reply:
the crux is not where` OSD::shutdown()` is called. `app_template::run()` stops the engine once the the future passed to it is returned. in our case, the future passed to it is but a bunch of `start()`s, so it has no knowledge of when the services stop. we need to to either wait for `OSD::stop()` in `app_template::run()` or continue using `app.run_deprecated()` and call `engine().exit()` explicitly. i'd prefer the second approach, see "crimson/osd: call engine().exit(0) after mkfs" in #26697 .